OAuth2 improvements (MBS-11058) + Content-Security-Policy

mwiencek opened the pull request:
If you're on the MusicBrainz team and have access to the OAuch Security Report, this should resolve all issues except TC4, TC11, TC12, and TC14 (some of those are breaking changes). This also sets a Content-Security-Policy on other pages that deal with account/admin data.

None of the changes here should be breaking except for the "Disallow same parameter twice" one, and that should be very low impact. The `Content-Security-Policy` headers might in theory interfere with userscripts and custom CSS in Firefox (I think Chrome is fine), *but* it's not set on any pages where userscripts typically run (edit and entity pages). We could see what impact it has after putting it on beta in any case.

I'm not sure that the report's analysis of our token entropy size was correct since we do use four random 32-bit integers from `Math::Random::Secure::irand` which is 128 bits. I did increase that to 192 bits anyway. So all OAuth-related tokens plus remember_me tokens will now be slightly longer.

reosarevok reviewed (commented):
In general it seems sensible to me (haven't tested) but see below plus typo "reponse_mode" on c87931a3b5126154465bf5180e8fd6f476809dfa message :)

reosarevok commented:
TC14 seems like a sensible thing to restrict, but I assume that would be a breaking change. Should we give notice and then block it in 6 months?

mwiencek commented on lib/MusicBrainz/Server/Controller/OAuth2.pm:
```diff
@@ -33,6 +33,9 @@ sub authorize : Local Args(0) RequireAuth SecureForm
     my ($self, $c) = @_;
 
+    # https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14#section-4.2.4
+    $c->res->header('Referrer-Policy' => 'origin-when-cross-origin');
```
That seems reasonable.

mwiencek commented on lib/MusicBrainz/Server/Data/EditorOAuthToken.pm:
```diff
@@ -42,10 +42,30 @@ sub _entity_class
+    my ($orig, $self, $row) = @_;
+
+    my $entity = $self->$orig($row);
+    # N.B. The authorization_code column stores the code_challenge and
+    # code_challenge_method, if supplied, separated by semi-colons.
+    my (
+        $authorization_code,
+        $code_challenge,
+        $code_challenge_method,
+    ) = split q(;), ($entity->authorization_code // '');
```
That's correct. All three codes are base64url-encoded. Besides letters and numbers, only `-` and `_` can appear there.

We could change it during the next schema change, yes. We *could* change it now if we're careful about it, even, by adding a `DBDefs` flag for enabling PKCE (disabled by default) that causes MBS to only reference the new table columns if enabled. Then we remove that flag in the schema change release. That might be preferable, actually.

mwiencek replied:
Done, see https://github.com/metabrainz/musicbrainz-server/pull/1692#issuecomment-692262040

mwiencek commented on t/lib/t/MusicBrainz/Server/Controller/OAuth2.pm:
```diff
@@ -17,12 +17,12 @@ sub oauth_redirect_ok
-    is($host, $uri->host);
-    is($path, $uri->path);
-    is($state, $uri->query_param('state'));
+    is($uri->scheme, 'http');
+    is($uri->host, $host);
+    is($uri->path, $path);
+    is($uri->query_param('state'), $state);
```
All tests. :) And it's not just a Perl thing, it's the same in any test framework I can think of besides JUnit, I guess. But the convention used depends on the test framework, which is the right place to find documentation:

https://metacpan.org/pod/Test::More#is
https://github.com/substack/tape#tequalactual-expected-msg

mwiencek commented on lib/MusicBrainz/Server/Controller/OAuth2.pm:
```diff
@@ -36,6 +36,7 @@ sub authorize : Local Args(0) RequireAuth CSRFToken
     # https://tools.ietf.org/html/draft-ietf-oauth-security-topics-14#section-4.14
     # Authorization servers MUST prevent clickjacking attacks.
     $c->response->header('X-Frame-Options' => 'DENY');
+    $c->response->header('Content-Security-Policy' => q(default-src 'self' staticbrainz.org));
```
Right, since I'm only setting it in `sub authorize` there. But I do set CSP headers on a ton more pages in the commit titled "Add CSP headers to other account forms." Which is mainly only on account/admin forms where CSRF tokens were added before, not edit pages.

mwiencek commented on t/lib/t/MusicBrainz/Server/Controller/OAuth2.pm:
```diff
@@ -464,13 +482,134 @@ test 'Exchange authorization code' => sub {
+    oauth_redirect_error(
+        $test->mech,
+        'www.example.com', '/callback', 'xyz',
+        'invalid_request', q(Invalid code_challenge; see https://tools.ietf.org/html/rfc7636#section-4.1),
```
Yup. It's unlikely the syntax highlighter supports Perl's quoting operators correctly, since you can use almost any delimiters you want - `q(foo) q[foo] q|foo| q;foo; q!foo! q^foo^ q<foo>` and more all work, and can be useful if you know none of those appear in the string. (`q` is equivalent to single quotes, for double quotes/variable substitution you'd use `qq`.)

Though I'm not sure why I didn't just use `''` on this line since the error message doesn't contain single quotes. I'll change that.

mwiencek commented:
> TC14 seems like a sensible thing to restrict, but I assume that would be a breaking change. Should we give notice and then block it in 6 months?

That makes sense though I'd like to first log to Sentry which applications are doing that so we can contact the application owners directly. Which I'd rather do in a followup PR. :)

mwiencek commented:
> TC12 is a breaking change, could be implemented as a simple warning to start with. TC11 and TC14 should probably be implemented in a compatible fashion for a transition period as @reosarevok suggested.

Re: TC11, that one in particular doesn't seem breaking to me, though the situations it applies in would be quite rare and we don't put any sensitive info in the fragment ourselves. It otherwise seems to be [annoying](https://www.npmjs.com/package/remove-fb-hash) to a [lot of people](https://stackoverflow.com/questions/7131909/facebook-callback-appends-to-return-url). But maybe it's a good idea regardless.

Re: TC12, I'm not sure we'll be implementing this. We delete the authorization code immediately when an access token is granted, so we can't detect if it's used more than once. (Obviously it's rejected on second use already, because we don't have it--but keeping it stored just to *detect* that doesn't necessarily seem like a good idea if it increases the chance of a bug where it can be reused.)

mwiencek commented:
Updated the PKCE commit with the necessary schema change. You have to set `OAUTH2_ENABLE_PKCE` to 1, otherwise we don't interact with that bit of the schema at all. So it's backwards-compatible with mirrors that don't have the schema change applied.

mwiencek commented:
I updated https://wiki.musicbrainz.org/Development/OAuth2 with details about PKCE/token revocation, so we can transclude that once it's released.

Also updated that to advise against including the `access_token` in the URI if possible.

reosarevok reviewed (approved):
LGTM

yvanzo reviewed (approved):
Double-checked changes for mirrors, look awesome now.
